src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+    const GraphTile* tile = reader.GetGraphTile(shortcut_id);
+    const DirectedEdge* shortcut = tile->directededge(shortcut_id);
+
+    // bail if this isnt a shortcut
+    if (!shortcut->is_shortcut()) {
+      return {shortcut_id};
+    }
+
+    // loop over the edges leaving its begin node and find the superseded edge
+    GraphId begin_node = reader.edge_startnode(shortcut_id);
+    if (!begin_node)
+      return {shortcut_id};
+
+    // loop over the edges leaving its begin node and find the superseded edge
+    std::vector<GraphId> edges;
+    for (const DirectedEdge& de : tile->GetDirectedEdges(begin_node.id())) {
+      if (shortcut->shortcut() & de.superseded()) {
+        edges.push_back(tile->header()->graphid());
+        edges.back().set_id(&de - tile->directededge(0));
+        break;
+      }
+    }
+
+    // bail if we couldnt find it
+    if (edges.empty()) {
+      LOG_TRACE("Unable to recover shortcut for edgeid " + std::to_string(shortcut_id) +
+                " | no superseded edge");
+      return {shortcut_id};
+    }
+
+    // seed the edge walking with the first edge
+    const DirectedEdge* current_edge = tile->directededge(edges.back());
+    uint32_t accumulated_length = current_edge->length();
+
+    // walk edges until we find the same ending node as the shortcut
+    while (current_edge->endnode() != shortcut->endnode()) {
+      // get the node at the end of the last edge we added
+      const NodeInfo* node = reader.GetEndNode(current_edge, tile);
+      if (!node)
+        return {shortcut_id};
+      auto node_index = node - tile->node(0);
+
+      // check the edges leaving this node to see if we can find the one that is part of the shortcut
+      current_edge = nullptr;
+      for (const DirectedEdge& edge : tile->GetDirectedEdges(node_index)) {
+        // are they the same enough that its part of the shortcut
+        // NOTE: this fails in about .05% of cases where there are two candidates and its not clear
+        // which edge is the right one. looking at shortcut builder its not obvious how this is
+        // possible as it seems to terminate a shortcut if more than one edge pair can be
+        // contracted... NOTE: because we change the speed of the edge in graph enhancer we cant use
+        // speed as a reliable determining factor
+        if (begin_node != edge.endnode() && !edge.is_shortcut() &&
+            edge.forwardaccess() == shortcut->forwardaccess() &&
+            edge.reverseaccess() == shortcut->reverseaccess() && edge.sign() == shortcut->sign() &&
+            edge.use() == shortcut->use() && edge.classification() == shortcut->classification() &&
+            edge.roundabout() == shortcut->roundabout() && edge.link() == shortcut->link() &&
+            edge.toll() == shortcut->toll() && edge.destonly() == shortcut->destonly() &&
+            edge.unpaved() == shortcut->unpaved() && edge.surface() == shortcut->surface()/* &&
+          edge.speed() == shortcut->speed()*/) {
```

Comment:
Forgotten comment?

Reply:
no that was there in the original. i left it unchanged. the comment above explains why its not turned on